subxt/src/backend/mod.rs

```diff
@@ -0,0 +1,351 @@
+// Copyright 2019-2023 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+//! This module exposes a backend trait for Subxt which allows us to get and set
+//! the necessary information (probably from a JSON-RPC API, but that's up to the
+//! implementation).
+
+pub mod legacy;
+pub mod rpc;
+
+use crate::error::Error;
+use crate::metadata::Metadata;
+use crate::Config;
+use async_trait::async_trait;
+use codec::{Decode, Encode};
+use futures::{Stream, StreamExt};
+use std::pin::Pin;
+use std::sync::Arc;
+
+/// Prevent the backend trait being implemented externally.
+#[doc(hidden)]
+pub(crate) mod sealed {
+    pub trait Sealed {}
```

Comment:
What is the reason to put `Sealed` into a separate `mod`? Would `pub(crate) trait Sealed {}` not suffice?

Reply:
Good question! This is a bit of a hack, because if you have `pub (crate) trait Sealed` on it's own then you run into:

```
crate-private trait `Sealed` in public interface
can't leak crate-private trait
```

But having the trait be fully `pub` (but in a module that people can't access, so the thing is pub but nobody can name it) circumvents this. It's weird, but it's a generally established pattern that crates use for this purpose :)

(I have only skimmed it but maybe eg https://predr.ag/blog/definitive-guide-to-sealed-traits-in-rust/ is interesting; it also talks about another approach via methods that can't be implemented :))

Reply:
Oh that is very interesting!